### PR TITLE
feat: passing callback to report progress

### LINKED ISSUE #190: Add a way to get progress report
I'd like to add callbacks to docker-compose functions (e.g. upAll, upMany) in order to get progress report.
For example:

```ts
 compose.upAll({ 
    cwd: path.join(__dirname),
    log: logOutput, 
    callback: (event) => console.log(event)  // this callback will fire before promise resolves
});


```
related discussion #188

### PR DESCRIPTION
**Pull request by upsilon01**
resolve #190

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -11,6 +11,7 @@ export interface IDockerComposeOptions {
   composeOptions?: string[] | (string | string[])[]
   commandOptions?: string[] | (string | string[])[]
   env?: NodeJS.ProcessEnv
+  callback?: (chunk: any) => void
```
Comment (Steveb-p) [inline src/index.ts:14]: I'm not exactly convinced that it should be done using callback function, but it looks as most simple solution.

However:
1. `chunk` should be better defined. Afaik it's an instance of `Buffer`, but it needs confirmation.
2. Second argument should be present to allow recognition from which stream (standard output or error) chunk is coming from.

```suggestion
  callback?: (chunk: Buffer, streamSource: 'stdout'|'stderr') => void
```

File: src/index.ts [hunk 2]
```diff
@@ -181,10 +182,12 @@ export const execCompose = (
 
     childProc.stdout.on('data', (chunk): void => {
       result.out += chunk.toString()
+      if(options.callback) options.callback(chunk)
```
Comment (Steveb-p) [inline src/index.ts:185]: [Optional chaining](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#optional-chaining) can be used instead.

```suggestion
      options.callback?.(chunk)
```

File: test/index.test.ts [hunk 3]
```diff
@@ -723,3 +723,24 @@ test('parse ps output', () => {
...
 })
+
+test('returns progres of command before promise resolves', async (): Promise<void> => {
+  
+  let firstCallbackFire
+
+  const config = {
+    cwd: path.join(__dirname),
+    config: './docker-compose.yml',
+    callback: (chunk) => {
```
Comment (Steveb-p) [inline test/index.test.ts:734]: This callback should properly define the types.

File: test/index.test.ts [hunk 4]
```diff
@@ -723,3 +723,24 @@ test('parse ps output', () => {
...
+      if (!firstCallbackFire && chunk.toString().includes('proxy')){
+        firstCallbackFire = Date.now()
+      } 
+    }
+  }
+
+  const resolveTime = await compose.upAll(config).then( () => Date.now() )
```
Comment (Steveb-p) [inline test/index.test.ts:741]: Equivalent to:
```suggestion
  await compose.upAll(config)
  const resolveTime = Date.now()
```
with the added benefit of not discarding the result of `compose.upAll`.

File: src/index.ts (hunk 1 above)
Comment (AlexZeitler) [reply src/index.ts:14]: @Steveb-p +1 for typings.

Comment (AlexZeitler) [reply src/index.ts:14]: I think we should start with minimal effort and watch the feedback.

Comment (upsilon01) [reply src/index.ts]: @Steveb-p  It is indeed a Buffer, fixed

File: test/index.test.ts (hunk 3 above)
Comment (upsilon01) [reply test/index.test.ts]: done

**Review by Steveb-p — APPROVED** [review]

**Review by AlexZeitler — CHANGES_REQUESTED** [review]
@upsilon01 The changes look good to me.
Would you mind updating the docs?

**Comment by upsilon01** [comment]
@AlexZeitler i updated the docs

**Review by AlexZeitler — APPROVED** [review]

**Comment by AlexZeitler** [comment]
@upsilon01 thanks for your contribution. 📦 `@0.23.14` is available on npm.

**Pull request merged**